controller/api/proxy/endpoints_watcher.go

```diff
@@ -243,6 +251,31 @@ func (e *endpointsWatcher) updateEndpoints(oldObj, newObj interface{}) {
 	e.addEndpoints(newObj)
 }
 
+func (e *endpointsWatcher) getState() servicePorts {
```

Comment:
What is this method doing that's different from just returning `e.servicePorts`?  Is it effectively making a deep copy of the map?  Is that necessary?

Reply:
oh, we copy the data out so that we can release the read mutex?  I think this could probably use a comment